pyiron_contrib/generic/s3io.py

```diff
@@ -0,0 +1,567 @@
+import fnmatch
+import io
+import json
+import os
+import posixpath
+from functools import lru_cache
+
+import boto3
+from botocore.client import Config
+
+from pyiron_base.generic.filedata import load_file, FileDataTemplate
+
+
+class S3FileData(FileDataTemplate):
+    """FileData stores an instance of a data file from an S3 system, e.g. a single Image from a measurement."""
+    def __init__(self, s3obj, filename=None, filetype=None):
+        """FileData class to store data and associated metadata.
+
+            Args:
+                s3obj (s3object): s3object containing a file with data
+                filename (str): filename associated with the data.
+                filetype (str): File extension associated with the type data,
+                                If provided this overwrites the assumption based on the extension of the filename.
+        """
+        self._s3obj = s3obj
+        if filename is None:
+            self.filename = s3obj.key.split('/')[-1]
+        else:
+            self.filename = filename
+        if filetype is None:
+            filetype = os.path.splitext(self.filename)[1]
+            if filetype == '' or filetype == '.':
+                self.filetype = None
+            else:
+                self.filetype = filetype[1:]
+        else:
+            self.filetype = filetype
+        self._data = None
+        self._metadata = self._s3obj.metadata
+
+    @property
+    @lru_cache()
+    def data(self):
+        """Return the associated data."""
+        if self._data is None:
+            self._s3obj = self._s3obj.get()
+            self._data = self._s3obj["Body"].read()
+        return load_file(io.BytesIO(self._data), filetype=self.filetype)
+
+    @property
+    def metadata(self):
+        return self._metadata
+
+    def __repr__(self):
+        return f"FileData containing {self.filename}."
+
+    def _ipython_display_(self):
+        result = f"File    {self.filename}   with meta data:\n"
+        result += "-------------------------------------------\n"
+        for key, value in self.metadata.items():
+            result += f"{key}:  {value}\n"
+        print(result)
+
+
+class S3ioConnect:
+    def __init__(self, credentials=None, bucket_name=None):
+        """Establishes connection to a specific 'bucket' of a S3 type object store.
+
+            Args:
+                credentials (str/dict/boto3.resource/None):
+                        if str: path to a json configuration file with login credentials for the bucket.
+                        if dict: dictionary containing the login credentials.
+                        if boto3.resource: This resource is used as is, requires bucket_name to be specified.
+                        if None: use Credentials as obtained by boto3, requires bucket_name to be specified.
+                bucket_name(str/None): Name of the bucket, overwrites name given in the config.
+
+            The configuration needs to provide the following information:
```

Comment:
```suggestion
            The credentials needs to provide the following information:
```
Was that a typo?  Otherwise I'm not sure where to pass additional keywords, since this method does not accept kwargs.